src/egui/mod.rs

```diff
@@ -277,6 +297,46 @@ impl UpdatesApp {
         }
     }
 
+    fn handle_merge_promises(&mut self, toasts: &mut Vec<(String, ToastLevel)>) {
+        for i in 0..self.v.merge_queue.len() {
+            let merge = &mut self.v.merge_queue[i];
+            if let Ok(status) = merge.progress_rx.try_recv() {
+                if let MergeStatus::PartProgress(progress) = status {
+                    merge.part_progress = progress;
+                }
+
+                merge.last_received_status = status;
+            }
+
+            let result = match merge.promise.ready() {
+                Some(res) => res,
+                None => continue
+            };
+
+            match result {
+                Ok(_) => {
+                    info!("Merge completed for {}", &merge.title_id);
+
+                    toasts.push((format!("{} merged successfully!", &merge.title_id), ToastLevel::Success));
+                    self.v.completed_merges.push(merge.title_id.clone());
+                }
+                Err(e) => {
+                    self.v.failed_merges.push(merge.title_id.clone());
+
+                    match e {
+                        MergeError::FilepathMismatch(_) | MergeError::PackagesUnmergable(_) | MergeError::FileMergeFailure => {
+                            toasts.push((format!("Failed to merge {}. Check the log for details.", merge.title_id), ToastLevel::Error));
+                        }
+                    }
+
+                    error!("Could not merge files for {}, reason: {:?}", merge.title_id, e);
+                }
+            }
+
+            self.v.merge_queue.remove(i);
```

Comment:
Nice catch, this is an obvious error on my part, I haven't noticed that I'm invalidating `i` by removing the element.
This error did not manifest during testing because to trigger it more than one merge has to finish on the same GUI repaint, and since those happen so often, it's rare to trigger such situation.

I guess borrow checker did not complain because while iterating the vector, the `merge` I'd borrowed for the loop was in scope but not used after the moment I borrowed mutably vector for removal - had I tried accessing the borrowed `merge` after borrowing the vector for removal, the borrow checker would intervene.

Reply:
It's a good thing I reviewed today instead of yesterday, it was easy to miss! I don't think it'd be super-easy to trigger, but it could lead to random errors/crashes which are always really fun to debug.